[1단계 - 출석 구현] 줄리(한승아) 미션 제출합니다.

jbilee opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - https://prolog.techcourse.co.kr/studylogs/3942


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

들여쓰기 깊이를 2까지 사용하고 getter도 불가피하게 사용하게 되어 4점을 선택했습니다.

  
## 어떤 부분에 집중하여 리뷰해야 할까요?

프로그램이 `오늘`을 판단하는 기준이 애플리케이션을 실행할 때 전달하는 프로그램 인수이기 때문에, 로컬 환경에서 프로그램을 실행시킬 때 "YYYY-MM-DD" 형식의 프로그램 인수를 전달해야 합니다. 출석 미션의 날짜 기준은 2024년 12월인데 `LocalDate.now()`는 2025년 2월 날짜를 반환하기에 미션과 동일한 환경에서 실행되는 애플리케이션을 시뮬레이팅하기 위해 이와 같이 구현했습니다.

크루별로 출결 기록을 `CrewAttendanceRecords` 내에 `Map<Crew, AttendanceRecords>`에 저장하였고, `AttendanceRecords`는 `List<AttendanceRecord>`를 가지는 일급 컬렉션으로 구현했습니다. 이로 인해 `CrewAttendanceRecords`와 `AttendanceRecords` 두 클래스에 `getAttendanceCount`나 `getSortedRecords` 등 중복되는 이름를 가진 메소드들을 생성하게 된 것 같은데 이를 더 추상화하는 것이 가능했을 지에 대한 고민이 들었습니다.

jbilee committed:
docs(readme): 기능 요구 명세 작성

- 애플리케이션 구현에 필요한 기능 목록 작성

jbilee committed:
chore: 출석 기록 .csv 파일 생성

jbilee committed:
feat(day, attendance): 출석 관리 기능 구현

- 입력한 출석 시간에 따라 출석, 지각 또는 결석으로 처리한다
- 등교 시간이 캠퍼스 운영 시간이 아닌 경우 예외를 발생시킨다

jbilee committed:
feat(disciplinary_status): 제적 위험자 관리 기능 구현

- 누적 지각 및 결석 횟수에 따라 제적 위험자로 관리한다

jbilee committed:
feat(attendance): 출석 확인 관련 기능 구현

- 등교일이 아닐 경우(주말, 공휴일) 예외를 발생시킨다

jbilee committed:
feat(attendance_record): 출석 관리 기능

- 프로그램 시작 시 출석 기록 파일에서 출석 기록을 불러온다.

jbilee committed:
feat(attendance_record): 출석 수정 기능

- 출석 기록이 없는 날짜를 입력한 경우 예외를 발생시킨다.

jbilee committed:
feat(attendance_record): 출석 수정 기능

- 출석 확인을 수정하려면 닉네임, 수정하려는 날짜, 등교 시간을 입력하여 기록을 수정할 수 있다.

jbilee committed:
feat(attendance_record): 출석 수정 기능

- 기록이 없는 닉네임을 입력하면 예외를 발생시킨다.

jbilee committed:
feat(attendance_record): 출석 기록 확인 기능

- 기록이 없는 닉네임을 입력하면 예외를 발생시킨다.

jbilee committed:
feat(attendance_record): 출석 확인 기능

- 기록이 없는 닉네임을 입력하면 예외를 발생시킨다.

jbilee committed:
feat(attendance_record): 출석 확인 기능

- 이미 출석을 하였는데 다시 출석 확인을 하는 경우 예외를 발생시킨다.

jbilee committed:
chore(global): 패키지 생성 및 파일 이동

- `domain` 패키지 생성
- 도메인에 해당하는 클래스를 `domain` 패키지로 이동
- `Application` 생성

jbilee committed:
feat(menu): 메뉴 기능

- 날짜와 메뉴 리스트를 안내한다.

jbilee committed:
feat(menu): 메뉴 기능

- 종료를 선택하면 프로그램을 종료한다.

jbilee committed:
feat(menu): 메뉴 기능

- 사용자가 입력한 값에 따라 해당 기능 실행한다.

jbilee committed:
feat(check_in): 출석 확인 기능

- 시간 입력 형식이 잘못된 경우 예외를 발생시킨다.

jbilee committed:
feat(check_in): 출석 확인 기능

- 닉네임과 등교 시간을 입력하면 출석할 수 있다.
- 출석 후 출석 기록을 확인할 수 있다.

jbilee committed:
feat(update_attendance): 출석 수정 기능

- 수정 후에는 변경 전과 변경 후의 출석 기록을 확인할 수 있다.
- 시간 입력 형식이 잘못된 경우 예외를 발생시킨다.

jbilee committed:
feat(check_attendance): 출석 기록 확인 기능

- 출석 기록 목록을 출력한다.

jbilee committed:
feat(check_attendance): 출석 기록 확인 기능

- 출결 현황을 출력한다. (출석, 지각, 결석)

jbilee committed:
feat(check_attendance): 출석 기록 확인 기능

- 닉네임을 입력하면 전날까지의 크루 출석 기록을 확인할 수 있다.
- 면담 대상자 여부를 출력한다.

jbilee committed:
feat(check_disciplinary): 제적 위험자 확인

- 전날까지의 크루 출석 기록을 바탕으로 제적 위험자를 파악한다.

jbilee committed:
feat(menu): 메뉴 기능

- 예외가 발생하면 초기 화면으로 돌아간다.

jbilee committed:
refactor(controller) : AttendanceController 리팩터링

- 메뉴 선택 로직 리팩터링
- 상수 추출
- 메서드 순서 변경

jbilee committed:
feat(update_attendance) : 출석 수정 기능

- 날짜 입력 형식이 잘못된 경우 예외를 발생시킨다.

jbilee committed:
refactor(view) : OutputView 리팩터링

- 메소드 순서 변경

jbilee committed:
refactor(domain) : Crew 리팩터링

- 레코드로 변경

jbilee committed:
refactor(domain) : Attendance 리팩터링

- 메소드 순서 변경
- enum 표기 방식 수정

jbilee committed:
refactor(day) : Day 리팩터링

- 사용하지 않는 isHoliday 멤버 변수 삭제
- getDay 메서드 실패시 throw 하도록 수정
- checkHoliday 로직 수정

jbilee committed:
refactor(domain) : DisciplinaryStatus 리팩터링

- 메소드 순서 변경
- 메소드 이름 변경

jbilee committed:
refactor(domain) : AttendanceRecord 리팩터링

- 정적 팩토리 메서드 사용
- 메서드 순서 변경

jbilee committed:
fix(day) : Day 버그 수정

- checkHoliday 메서드 오류 수정

jbilee committed:
refactor(domain) : 중복된 메서드 통합

- getAttendanceCount 메서드 생성

jbilee committed:
refactor(domain) : 메서드 분리

- getDisciplinaryStatus 메서드 생성

jbilee committed:
style(attendance_records) : 스타일 통일

- 줄바꿈 추가

jbilee committed:
refactor(domain) : CrewAttendanceRecords 생성 방식 리팩터링

- 팩토리 메서드 패턴 사용
- CsvParsingGenerator 추가

jbilee committed:
refactor(domain) : CrewAttendanceRecords 리팩터링

- 메서드 분리
- 메서드 순서 변경

jbilee committed:
refactor(global) : 전체적인 코드 리팩터링

- 미사용 메서드 삭제
- 생략된 final 키워드 추가
- 가독성을 위한 라인 분리

jbilee committed:
refactor(domain) : CsvParsingGenerator 리팩터링

- 불필요한 매개변수 삭제

ddaaac commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,83 @@
+            } catch (DateTimeParseException e) {
+                throw new IllegalArgumentException("[ERROR] 프로그램 인수를 YYYY-MM-DD 형식으로 입력해 주세요.");
+            }
+        };
```
> 프로그램이 오늘을 판단하는 기준이 애플리케이션을 실행할 때 전달하는 프로그램 인수이기 때문에, 로컬 환경에서 프로그램을 실행시킬 때 "YYYY-MM-DD" 형식의 프로그램 인수를 전달해야 합니다. 출석 미션의 날짜 기준은 2024년 12월인데 LocalDate.now()는 2025년 2월 날짜를 반환하기에 미션과 동일한 환경에서 실행되는 애플리케이션을 시뮬레이팅하기 위해 이와 같이 구현했습니다.

날짜를 주입하려고 하셨군요. 궁금한 점인데 람다로 만든 이유가 있나요? 그냥 `currentDate = LocalDate.parse(args)` 처럼 구현할 수 있지 않나 싶어서요.

jbilee replied:
코드를 테스트할 때 외부에서 오늘 날짜를 원하는 값으로 제어할 수 있기 위해, 그리고 잘못된 프로그램 인수로 인해 날짜가 파싱이 안되는 경우를 catch해서 커스텀 오류 메세지를 출력하기 위해 람다식으로 구현했는데 범블비의 말씀을 듣고 다시 생각해보니 generator 인터페이스를 두지 않고 `LocalDate`으로 저장해도 원하는 만큼 제어가 가능할 것 같습니다. 또한 람다식 내부에 있는 try-catch 블럭은 `Application`에 구현해도 충분하지 않을까 싶습니다.

jbilee replied:
[refactor(application): 현재 날짜 구현 방식 수정](https://github.com/woowacourse/java-attendance/pull/44/commits/2ca17d60064f120f973f92a429022a9b0feb590d)

ddaaac commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,83 @@
+        do {
+            menuInput = retryUntilSuccess(() -> {
+                String input = inputView.readMenu(currentDateGenerator.generate());
+                menu.get(input).run();
```
input에 해당하는 값이 없는 경우도 처리해주면 좋을 것 같네요,.

jbilee replied:
사전에 `input` 값을 검증해서 메뉴 옵션 중 하나가 아니라면 throw하도록 구현해서 해당하는 값이 없음으로 인해 발생하는 문제는 없을 거라고 생각했는데, 직접 애플리케이션을 다시 돌려보니 `input` 값이 "Q"일 때에 대한 처리를 누락시켜서 NPE 예외가 발생한다는 점을 발견했습니다 😭 말씀 주신 대로 값이 없는 경우까지 처리해서 전체적으로 오류를 핸들링할 수 있도록 수정하려고 합니다!

jbilee replied:
[refactor(controller): 컨트롤러 내부 로직 리팩터링](https://github.com/woowacourse/java-attendance/pull/44/commits/c89d74a5301d3894cfb610742b1b5b95073a52e6)

ddaaac commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,83 @@
+        this.currentDateGenerator = () -> {
+            try {
+                return LocalDate.parse(args[START_DATE_INDEX]);
+            } catch (DateTimeParseException e) {
```
인수 자체가 없을 경우 어떻게 예외 처리가 될까요?

jbilee replied:
인수 자체가 없으면 `ArrayIndexOutOfBoundsException` 예외가 발생하는데, 이는 현재 컨트롤러 생성자에서는 따로 캐치하고 있지 않다는 것을 깨달았습니다.. 캐치 블럭을 추가해 LocalDate 파싱이 실패할 경우와 동일한 오류 메세지를 가진 `IllegalArgumentException` 예외를 다시 throw하려고 합니다.

jbilee replied:
[refactor(controller): 컨트롤러 리팩터링](https://github.com/woowacourse/java-attendance/pull/44/commits/2188de724677f4e4d6de9b8bff0b8f8a81cc9c4d)

ddaaac commented on src/main/java/domain/CrewAttendanceRecords.java:
```diff
@@ -0,0 +1,116 @@
+        validateCrewPresence(crew);
+        AttendanceRecords records = crewAttendanceRecords.get(crew);
+        records.addRecord(newAttendanceRecord);
+        return records.removeRecord(newAttendanceRecord.getDate());
```
여기서는 새 날짜를 추가하고 -> 지우는 순서인데
보통은 지우고 -> 넣는 방식을 사용합니다.

후자는 지울 때 어떤 식으로 지워도 문제가 없는데 전자는 컬렉션의 형태나 찾는 방법에 따라 버그가 발생할 수 있거든요.

jbilee replied:
지우기 전 새 데이터를 먼저 추가함으로 인해 발생할 수 있는 문제는 미처 고려하지 못했습니다! 이 부분 수정하도록 하겠습니다. 🙂

jbilee replied:
[refactor(attendance_update): 출석 기록 수정 기능 리팩터링](https://github.com/woowacourse/java-attendance/pull/44/commits/cc5c1397a3ab52e1017ac6b7017c1a8a42c3f1ce)

ddaaac commented on src/main/java/domain/CrewAttendanceRecords.java:
```diff
@@ -0,0 +1,116 @@
+    public AttendanceRecord updateAttendanceRecord(Crew crew, AttendanceRecord newAttendanceRecord) {
+        validateCrewPresence(crew);
+        AttendanceRecords records = crewAttendanceRecords.get(crew);
+        records.addRecord(newAttendanceRecord);
```
updateAttendanceRecord란 메서드 이름만 보고 이 메서드의 반환값이 이전 레코드란 걸 알 수 있을까요?

jbilee replied:
`remove`와 같은 메소드가 리스트에서 요소를 제거한 뒤에 제거된 해당 요소를 결과로 반환해주기에 `updateAttendanceRecord` 메소드도 동일한 맥락으로 생각했었는데, 외부인의 입장을 고려해봤을 때 사실 메소드의 이름만 봐서는 알 수 없을 것 같습니다. 보다 이해하기 쉬운 흐름을 고안해보겠습니다.

jbilee replied:
[refactor(attendance_update): 출석 기록 수정 기능 리팩터링](https://github.com/woowacourse/java-attendance/pull/44/commits/cc5c1397a3ab52e1017ac6b7017c1a8a42c3f1ce)

ddaaac commented on src/main/java/domain/AttendanceRecords.java:
```diff
@@ -0,0 +1,72 @@
+public class AttendanceRecords {
+    private static final LocalDate FILL_START_DATE = LocalDate.of(2024, 11, 30);
+
+    private final List<AttendanceRecord> attendanceRecords;
```
사소하지만 어차피 sort해서 보여줄거라면 Set을 써도 될 것 같네요.

jbilee replied:
지금까지 Set(HashSet)은 정렬할 수 없는 컬렉션이라고 알고 있었어서 피드백 주신 내용을 이해하기 위해 알아봤고, 새로운 값을 추가하면 자동으로 정렬되어 저장되는 TreeSet의 존재를 알게 됐습니다! 리팩터링하면서 TreeSet으로 적용해보려고 합니다.

ddaaac replied:
이전 코드에서 List<>에 값을 빼고 넣을 때 순서가 바뀌어서 남겼던 코멘트였어요
TreeSet 잘 활용해주셨네요 👍

jbilee replied:
[refactor(attendance_records): AttendanceRecords 클래스 리팩터링](https://github.com/woowacourse/java-attendance/pull/44/commits/d50099a7b02d4c85aebf7f5d1de0aea70f1d30c4)

ddaaac commented on src/main/java/domain/Attendance.java:
```diff
@@ -0,0 +1,44 @@
+        validateOpenTime(time);
+        LocalTime start = day.getStart();
+        Duration between = Duration.between(start, time);
+        if (between.compareTo(TARDY.thresholdInMinutes) < 0) {
```
값을 꺼내서 로직을 구현하는 대신 Attendance 객체에게 메세지를 던져보는 건 어떨까요?

jbilee replied:
`PRESENT`, `TARDY`, `ABSENT`도 각각의 `Attendance` 객체이기 때문에 위와 같이 구현하면 `TARDY` 인스턴스의 `thresholdInMinutes`를 호출해 직접 값을 꺼내서 쓰게 되는 것으로 이해했습니다..! 따라서 출석 시간을 주면 본인이 어떤 상태에 해당하는지를 상태 목록 중에서 필터링해서 찾아달라는 식으로 수정하려고 합니다.

jbilee replied:
[refactor(attendance): Attendance 리팩터링](https://github.com/woowacourse/java-attendance/pull/44/commits/088ad4f55398014c914dceeaa9ac09f5390dc161)

ddaaac commented on src/main/java/domain/AttendanceRecords.java:
```diff
@@ -0,0 +1,72 @@
+        for (LocalDate date = dateGenerator.generate().minusDays(1); date.isAfter(FILL_START_DATE); date = date.minusDays(1)) {
+            if (!hasRecordOfDate(date) && !Day.checkHoliday(date)) {
+                this.attendanceRecords.add(AttendanceRecord.asAbsent(date));
+            }
```
이런 부분은 메서드 분리를 해서 depth를 줄일 수 있어보여요.

jbilee replied:
이 부분 개선해보겠습니다, 피드백 감사합니다! 아직 메소드 분리를 어디까지 해야 적절한지에 대한 감이 잘 오지 않는 것 같습니다 😅

jbilee replied:
[refactor(attendance_records): AttendanceRecords 클래스 리팩터링](https://github.com/woowacourse/java-attendance/pull/44/commits/fdd97de806796ff2e524c1ec5e2909253909e0bc)

ddaaac commented on src/main/java/domain/AttendanceRecord.java:
```diff
@@ -0,0 +1,80 @@
+    public static final LocalTime ABSENT_TIME = LocalTime.of(14, 0);
+
+    private final LocalDate date;
+    private final LocalTime time;
```
LocalDateTime을 알아봐도 좋을 것 같아요.

jbilee replied:
로직 중에 날짜만 필요하거나 시간만 필요한 경우가 있어서(시간만 수정하거나 csv 파일에 기록이 없어서 날짜만 입력받아 출결 기록을 생성해야 하는 경우 등), 미리 분리해두면 `toLocalDate()`나 `toLocalTime()` 메소드를 추가적으로 호출할 필요가 없을테니 더 용이하겠다고 생각했었습니다. 😯

하지만 지금 드는 생각이, LocalDate/LocalTime으로 변환하는 케이스는 의외로 많지 않고 오히려 LocalDateTime을 사용하면 날짜와 시간을 원하는 형태로 한꺼번에 출력할 수 있기 때문에 불필요한 포맷팅 단계를 줄일 수 있을 것 같습니다!

jbilee replied:
[refactor(attendance_record): AttendanceRecord 클래스 리팩터링](https://github.com/woowacourse/java-attendance/pull/44/commits/6c77d7ff77958baea2342c208f908a265cb140b9)

ddaaac commented on src/main/java/domain/CrewAttendanceRecords.java:
```diff
@@ -0,0 +1,116 @@
+        return crewAttendanceRecords.get(crew).getAbsentCount();
+    }
+
+    public int getAttendanceCount(Crew crew, Attendance attendance) {
+        validateCrewPresence(crew);
+        return crewAttendanceRecords.get(crew).getAttendanceCount(attendance);
+    }
```
> 이로 인해 CrewAttendanceRecords와 AttendanceRecords 두 클래스에 getAttendanceCount나 getSortedRecords 등 중복되는 이름를 가진 메소드들을 생성하게 된 것 같은데 이를 더 추상화하는 것이 가능했을 지에 대한 고민이 들었습니다.

이런 건 중복보다는 각 객체의 책임 관점에서 접근해보면 좋습니다.

CrewAttendance, Attendance Records 두 객체 모두 결석 횟수를 구하는 책임을 가져야하는 게 적당해보입니다. CrewAttendanceRecords는 이를 더 잘 처리할 수 있는 AttendanceRecords에게 물어보는 거죠.
둘 모두 적당한 책임을 갖고 있으니 문제 없어보여요.

jbilee replied:
출석 미션 만큼의 규모를 가진 애플리케이션을 코딩해본 경험이 적다보니 `CrewAttendance`와 `AttendanceRecords`가 어딘가 비슷해 보이는 책임을 가지는게 어색하게 느껴졌던 것 같습니다. 피드백 감사합니다!

ddaaac commented on src/main/java/domain/CrewAttendanceRecords.java:
```diff
@@ -0,0 +1,116 @@
+    private List<Crew> sortWarnedCrews(List<Crew> crews) {
+        sortCrewsByName(crews);
+        sortCrewsByAttendanceCount(crews);
+        sortCrewsByDisciplinaryStatus(crews);
```
이렇게 정렬하면 마지막 정렬기준만 적용되는 거 아닐까요?
테스트도 보완하면 좋겠네요.

jbilee replied:
테스트 시엔 예상한 대로 정렬돼서 문제가 없는 줄 알았는데 한가지의 케이스만 확인해서 우연찮게 테스트가 성공했던 것 같습니다! 테스트와 함께 개선해보겠습니다.

jbilee replied:
[fix(disciplinary_status): 제적 위험자 관련 로직 버그 수정](https://github.com/woowacourse/java-attendance/pull/44/commits/c2f4bcf2875a3e7f6579e3ea454155f3242a0fb9)

[test(attendance_records): CrewAttendanceRecords 테스트 수정](https://github.com/woowacourse/java-attendance/pull/44/commits/72dc20a52dda155e31e378938ad30436fef11a6c)

ddaaac commented on src/main/java/domain/CrewAttendanceRecords.java:
```diff
@@ -0,0 +1,116 @@
+            int absentCount = crewAttendanceRecords.get(crew).getAbsentCount();
+            int tardyCount = crewAttendanceRecords.get(crew).getTardyCount();
+            absentCount += (tardyCount / 3);
+            return (absentCount + tardyCount % 3) * -1;
```
비슷한 로직이 다른 클래스에도 있던데, 코드만 보고 쉽게 파악하기 힘든 로직인만큼 중복을 제거하면서 로직을 한 군데에서 관리할 수 있으면 좋을 것 같아요.

jbilee replied:
지각 3회는 결석 1회로 계산한다는 요구사항으로 인해 한눈에 파악하기 어려운 로직이 나온 것 같습니다. (실제로 계산식을 만드는 과정에서도 어려움을 겪었습니다..) 중복을 줄이고 책임이 더 적합한 곳에서 결석을 계산하도록 개선해보겠습니다.

jbilee replied:
[fix(disciplinary_status): 제적 위험자 관련 로직 버그 수정](https://github.com/woowacourse/java-attendance/pull/44/commits/c2f4bcf2875a3e7f6579e3ea454155f3242a0fb9)

ddaaac reviewed (changes requested):
안녕하세요 줄리 👋 

복잡한 미션 구현 잘 해주셨네요.

몇 가지 코멘트 남겼으니 확인해주세요.

jbilee committed:
refactor(domain): 미사용 클래스 삭제

- `CurrentDateGenerator` 클래스 삭제

jbilee committed:
refactor(controller): 컨트롤러 내부 로직 리팩터링

- 메뉴옵션 Map에 종료에 대한 Runnable 추가
- 메뉴옵션 Map에 존재하지 않는 옵션으로 인해 발생하는 NullPointerException을 캐치하도록 수정
- 뷰에서 메뉴옵션 input을 대문자로 반환하도록 수정

jbilee committed:
refactor(attendance_records): AttendanceRecords 클래스 리팩터링

- Depth를 줄이기 위해 메소드 분리

jbilee committed:
refactor(attendance_record): AttendanceRecord 클래스 리팩터링

- LocalDate과 LocalTime의 상태를 분리해서 저장하던 부분을 하나의 LocalDateTime으로 저장

jbilee committed:
refactor(attendance_records): AttendanceRecords 클래스 리팩터링

- AttendanceRecord를 List 대신 TreeSet으로 저장하도록 수정
- AttendanceRecord 목록을 정렬하는 로직 삭제

jbilee committed:
refactor(attendance_update): 출석 기록 수정 기능 리팩터링

- 출석 기록 수정 시 기존 기록을 먼저 삭제하도록 수정
- 기록 수정 메소드가 기존 기록을 반환하지 않도록 수정

jbilee committed:
refactor(controller): 컨트롤러 리팩터링

- 프로그램 인수가 없는 경우에 대한 예외 처리 추가

jbilee committed:
test(attendance_records): 테스트 코드 수정

- `removeRecordTest` 테스트의 assertion 방법 변경
- 출석 기록이 없는 날짜 삭제에 대한 테스트를 출석 기록이 없는 날짜를 get할 시의 테스트로 변경
- 정렬 기능 테스트를 저장 기능 테스트로 변경

jbilee committed:
fix(attendance_records): CrewAttendanceRecords 클래스 버그 수정

- `getRecordAtDate` 메소드에서 크루 검증 단계를 누락한 부분 수정

jbilee committed:
test(attendance_records): CrewAttendanceRecords 테스트 수정

- 비즈니스 로직 리팩터링에 따른 테스트 코드 수정
- 정렬 기능 삭제에 따라 관련 테스트 삭제

jbilee committed:
refactor(attendance): Attendance 리팩터링

- Attendance 내부적으로 제적 상태를 반환하도록 수정

jbilee committed:
fix(disciplinary_status): 제적 위험자 관련 로직 버그 수정

- 제적 위험자를 정렬하는 로직 중 마지막 정렬 기능만 적용하던 이슈 수정
- 지각을 결석으로 변환하는 로직을 `DisciplinaryStatus`에서만 처리하도록 수정
- 미사용 메소드 삭제

jbilee committed:
test(attendance_records): CrewAttendanceRecords 테스트 수정

- 정확한 테스트를 위해 테스트 시 사용할 값을 주입하도록 수정

jbilee committed:
refactor(application): 현재 날짜 구현 방식 수정

- `DateGenerator`를 사용하는 대신 현재 날짜를 LocalDate로 저장하도록 수정
- `DateGenerator` 인터페이스 삭제

jbilee committed:
test(attendance): 출결 관련 클래스 테스트 수정

- `DateGenerator` 인터페이스가 삭제됨에 따라 테스트 내에서 현재 날짜를 주입하는 방식 수정

jbilee committed:
refactor(attendance_record): 출결 기록 관련 클래스 리팩터링

- 불필요한 생성자 삭제
- `Comparable`을 원시 타입으로 사용하던 부분에 타입 명시

jbilee committed:
test(attendance_records): `AttendanceRecords` 테스트 코드 수정

- 정확한 객체 비교를 통해 결석 기록에 대한 기능을 테스트하도록 수정

jbilee committed:
refactor(attendance_records): `CrewAttendanceRecords` 클래스 리팩터링

- 테스트에만 사용되던 메소드 삭제
- 삭제된 메소드에 대한 테스트를 `getRecordAtDate`에 대한 테스트로 수정

ddaaac commented on src/main/java/domain/CrewAttendanceRecords.java:
```diff
@@ -0,0 +1,103 @@
+    private final Map<Crew, AttendanceRecords> crewAttendanceRecords;
+
+    public CrewAttendanceRecords(CrewAttendanceRecordsGenerator generator, LocalDate currentDate) {
+        this.crewAttendanceRecords = generator.generate(currentDate);
```
의존성에 대한 고민도 가볍게 해볼 수 있겠네요.
CrewAttendanceRecords가 CrewAttendanceRecordsGenerator를 아는 것보다는 반대 의존성이 조금 더 자연스럽게 느껴지는데 줄리 생각은 어떠신가요?

jbilee replied:
말씀하신 반대된 의존성의 상황이 잘 떠오르지 않아서 아래와 같이 생각을 정리해봤습니다:
- 현재 `CrewAttendanceRecords`가 `CrewAttendanceRecordsGenerator`를 알아야 하는 이유는 프로그램 시작 시 csv 파일로 생성한 객체 데이터를 받아 저장해야 하기 때문
- 위 스텝 이후에는 `CrewAttendanceRecordsGenerator`와 `CrewAttendanceRecords`는 아무런 연관성이 없으니, 가능하다면 `CrewAttendanceRecords`가 `CrewAttendanceRecordsGenerator`의 존재를 모르게 하면 더 독립적인 객체가 될 수 있을 것 같음 (지금은 데이터를 받기 위해 `generator.generate()` 메소드를 `CrewAttendanceRecords` 생성자 안에서 호출한 것처럼, generator가 어떻게(=어떤 메소드를 통해) 데이터를 생성하는지를 알아야 하는 상태라서 외부에 의존적인 객체라고 생각됨)
- 애초에 `CrewAttendanceRecordsGenerator`가 Map<Crew, AttendanceRecords>를 반환하니까 `CrewAttendanceRecords`한테 데이터 생성 이후에는 필요없는 `CrewAttendanceRecordsGenerator`와 LocalDate을 전달해서 `CrewAttendanceRecords` 내부적으로 값을 만들도록 하는 대신, 생성자에 `CrewAttendanceRecordsGenerator`가 반환한 Map<Crew, AttendanceRecords>만 전달해주는 방식이 더 자연스러운 플로우겠다고 생각

고민해보니 `CrewAttendanceRecords`에 `CrewAttendanceRecordsGenerator`를 전달하는 대신 `CrewAttendanceRecordsGenerator`가 `CrewAttendanceRecords` 객체를 반환하는 것이 더 이해하기 쉬운 흐름일 것 같아서 이 부분을 수정해보려고 합니다!

jbilee replied:
[refactor(attendance_records): CrewAttendanceRecords 객체 생성 방식 변경](https://github.com/woowacourse/java-attendance/pull/44/commits/bebb2efd9248b383d3a5397ba27f6b4e7414f887)

ddaaac replied:
생각 잘 정리해주셨네요. 
CrewAttendanceRecords의 생성자도 훨씬 깔끔해졌어요!

ddaaac commented on src/main/java/domain/AttendanceRecords.java:
```diff
@@ -0,0 +1,75 @@
+                .orElseThrow(() -> new IllegalArgumentException("[ERROR] 출석 기록이 없는 날짜는 수정할 수 없습니다.\n"));
+    }
+
+    public void fillAbsences(LocalDate currentDate) {
```
테스트도 작성해주세요!

jbilee replied:
누락된 테스트 추가 완료했습니다!

[test(attendance_records): AttendanceRecords 클래스 테스트 추가](https://github.com/woowacourse/java-attendance/pull/44/commits/8f49250c121eab5e1a1fe31ee67d7242d4a6327d)

ddaaac commented on src/main/java/domain/DisciplinaryStatus.java:
```diff
@@ -0,0 +1,39 @@
+    NONE("해당 사항 없음", 2),
+    WARNING("경고", 3),
+    ONE_ON_ONE("면담", 6),
+    EXPELLED("제적", 31);
```
요구사항을 보면 

> 경고 대상자: 결석 2회 이상

> 면담 대상자: 결석 3회 이상

> 제적 대상자: 결석 5회 초과


으로 되어있습니다.

이 요구사항과 코드가 조금 더 시각적으로 연관이 되면 좋을 것 같아요.

jbilee replied:
확실히 요구사항에 나온 값들을 상수와 맞추는 것이 이해하기 쉬울 것 같습니다! 이 부분 수정해서 반영하겠습니다. :)
(조건에 이상과 초과가 섞여있는데 다수결을 따라 이상을 기준으로 잡으려고 합니다..)

jbilee replied:
[refactor(disciplinary_status): DisciplinaryStatus 클래스 리팩터링](https://github.com/woowacourse/java-attendance/pull/44/commits/3c77e9086110ed60d6076efb2ef6aa3e7f7d0b4e)

ddaaac commented on src/main/java/domain/CrewAttendanceRecords.java:
```diff
@@ -0,0 +1,103 @@
+        crews.sort(Comparator.comparing((Crew crew) -> {
+            int absentCount = crewAttendanceRecords.get(crew).getAbsentCount();
+            int tardyCount = crewAttendanceRecords.get(crew).getTardyCount();
+            int convertedAbsencesAndTardies = DisciplinaryStatus.getConvertedAbsencesAndTardies(absentCount, tardyCount);
```
get해서 계산하기보다는 `AttendanceRecords`에 메세지를 던져보면 좋을 것 같아요.

jbilee replied:
결국엔 `AttendanceRecords`만 지각과 결석 횟수를 구할 수 있으니, `absentCount`와 `tardyCount`를 따로 가져오지 않고 `AttendanceRecords`가 해당 횟수들을 가지고 스스로 결론을 반환하도록 해야 하는 것으로 이해했습니다..! 메세지를 던진다는 부분 유념하면서 수정하겠습니다.

jbilee replied:
[refactor(attendance_records): CrewAttendanceRecords 클래스 로직 리팩터링](https://github.com/woowacourse/java-attendance/pull/44/commits/7ea84a6ebeec0d16b8263125759afd74fe63b1a8)

[test(attendance_records): AttendanceRecords 테스트 추가](https://github.com/woowacourse/java-attendance/pull/44/commits/c441d15ac22491835a069a6a9dc61af363d08155)

ddaaac commented on src/main/java/domain/CrewAttendanceRecords.java:
```diff
@@ -0,0 +1,103 @@
+    public List<Crew> getWarnedCrews() {
+        List<Crew> expelledCrews = getCrewsByDisciplinaryStatus(DisciplinaryStatus.EXPELLED);
+        List<Crew> oneOnOneCrews = getCrewsByDisciplinaryStatus(DisciplinaryStatus.ONE_ON_ONE);
+        List<Crew> warningCrews = getCrewsByDisciplinaryStatus(DisciplinaryStatus.WARNING);
```
구현하기 나름이긴 할텐데, 제적->면담->경고 순 출력도 아래 sort 로직에 넣는 것도 좋아보여요.
어떤 순으로 정렬되는가를 확인할 때 빠르게 파악할 수 있을 것 같다는 생각도 드네요.

jbilee replied:
처음에는 제적 상태를 카테고리화해서 제적, 면담, 경고를 각각의 블럭이라는 개념으로 생각했는데, 결국엔 다같은 정렬에 대한 로직이기 때문에 `sortWarnedCrews`에서 제적 상태에 따른 정렬도 처리하는게 더 적절한 것 같습니다. 이 부분은 enum을 비교하는 방법을 더 학습해서 개선해보겠습니다!

jbilee replied:
[refactor(attendance_records): 제적 위험자 정렬 로직 수정](https://github.com/woowacourse/java-attendance/pull/44/commits/88a04d62320b1f1ad2cb245ad0b9f1d4729a5c30)

ddaaac commented on src/main/java/view/OutputView.java:
```diff
@@ -0,0 +1,82 @@
+
+    private String getDisplayTime(LocalTime time, Attendance attendance) {
+        if (attendance == Attendance.ABSENT) {
+            return "--:--";
```
View랑 도메인 분리 잘 해주셨네요 👍

ddaaac commented on src/main/java/domain/Attendance.java:
```diff
@@ -0,0 +1,46 @@
+    public static Attendance getAttendanceStatus(Day day, LocalTime time) {
+        validateOpenTime(time);
+        LocalTime start = day.getStart();
+        Duration duration = Duration.between(start, time);
```
변수명을 "늦은 정도" 로 바꿔도 좋을 것 같네요.

jbilee replied:
변수명을 `lateness` 혹은 `latenessThreshold`로 생각하고 있는데 이미 해당 변수를 전달하는 `isWithinThreshold` 메소드에 threshold라는 단어가 들어가서 `lateness`로 수정할까 합니다..!

jbilee replied:
[refactor(attendance): Attendance 클래스 리팩터링](https://github.com/woowacourse/java-attendance/pull/44/commits/1e0236cbf11ef4bb28550683d666b3125c69b8f6)

ddaaac commented on src/main/java/domain/Attendance.java:
```diff
@@ -0,0 +1,46 @@
+
+    private static void validateOpenTime(LocalTime time) {
+        if (time.isBefore(OPEN_HOUR) || time.isAfter(CLOSE_HOUR)) {
+            throw new IllegalArgumentException("[ERROR] 캠퍼스 운영 시간에만 출석이 가능합니다.\n");
```
에러 메세지도 고민해보면 좋을 것 같아요.

이 validate이 불리는 컨텍스트는 출석 체크를 할 때이긴 하지만

나중에 다른 로직에서 이 검증이 불릴 수도 있습니다.

이 객체만 놓고 봤을 때 좀 더 자연스러운 메세지를 사용해도 좋을 것 같네요.

jbilee replied:
말씀주신 대로 출석이라는 특수한 상황을 배제하고 다른 상황에서 validate이 쓰일 경우를 상상해봤고, 이에 따라 단순히 `OPEN_HOUR`, `CLOSE_HOUR`만 고려해서 에러 메세지를 "운영 시간에 해당하지 않습니다."로 수정하려고 합니다.

jbilee replied:
[refactor(attendance): Attendance 검증 예외 메세지 리팩터링](https://github.com/woowacourse/java-attendance/pull/44/commits/48120c1b4c8e84329fdfe2fbc0074e3a9472a387)

[test(attendance): Attendance 테스트 코드 수정](https://github.com/woowacourse/java-attendance/pull/44/commits/b96aa0f9b35b974d58f0b76e172afd9a5adbffa4)

ddaaac reviewed (changes requested):
안녕하세요 쥴리 👋 

구현 너무 깔끔하게 해주셨네요.

다음에 요청하시면 머지할 것 같아요! 추가로 궁금하신 점 있으면 남겨주세요.

jbilee committed:
chore(test): 테스트 코드 이동

- `AttendanceRecordsTest`에 있던 `CrewAttendanceRecords` 생성과 관련된 테스트 코드를 `CrewAttendanceRecordsTest`로 이동

jbilee committed:
test(attendance_records): `AttendanceRecords` 클래스 테스트 추가

- `fillAbsences` 메소드 구현에 대한 테스트 작성

jbilee committed:
refactor(disciplinary_status): `DisciplinaryStatus` 클래스 리팩터링

- 요구사항과 시각적으로 일치하도록 상수값 변경

jbilee committed:
refactor(attendance): `Attendance` 클래스 리팩터링

- 변수명을 직관적으로 수정

jbilee committed:
refactor(attendance): `Attendance` 검증 예외 메세지 리팩터링

- `validateOpenTime`에 대한 에러 메세지를 `Attendance` 객체의 관점에서 적절한 내용으로 수정

jbilee committed:
refactor(attendance_records): `CrewAttendanceRecords` 클래스 로직 리팩터링

- 변환된 후의 총 결석과 지각 횟수를 `AttendanceRecords`가 직접 계산해 반환하도록 수정

jbilee committed:
test(attendance_records): `AttendanceRecords` 테스트 추가

- 추가된 메소드(`getConvertedAbsencesAndTardies`)에 대한 테스트 작성

jbilee committed:
test(attendance): `Attendance` 테스트 코드 수정

- 변경된 에러 메세지를 확인하도록 테스트 코드 수정

jbilee committed:
refactor(attendance_records): 제적 위험자 정렬 로직 수정

- 제적 위험자를 정렬하는 로직을 한 메소드에서 전부 수행하도록 수정

jbilee committed:
refactor(attendance_records): `CrewAttendanceRecords` 객체 생성 방식 변경

- `CrewAttendanceRecords` 클래스에서 `CrewAttendanceRecordsGenerator`에 대한 의존성 삭제
- `CrewAttendanceRecordsGenerator` 구현체가 직접 `CrewAttendanceRecords` 인스턴스를 생성하도록 수정

ddaaac commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    public void updateAttendance() {
+        Crew crew = inputView.readUpdateNickname();
+        LocalDate date = inputView.readUpdateDate();
+        AttendanceRecord oldRecord = crewAttendanceRecords.getRecordAtDate(crew, date);
+        LocalTime time = inputView.readUpdateTime();
```
정말 사소한 거긴 한데
input을 먼저 다 받고 로직 진행해도 좋을 것 같아요

취향차가 조금 있겠지만 쭉 읽어나갈 때 입력 받는 부분과 로직 진행 부분이 분리되어있는 게 좀 더 편한 거 같네요.

jbilee replied:
이것과 관련해서 들었던 고민이, 만약 먼저 받는 input이 유효하지 않을 경우 다음 input을 받기 전에 미리 예외를 던져야 하지 않을까 싶었습니다 😯 읽기 "좋은 코드"라는 것을 생각해보면 로직을 마지막에 처리해서 분리하는 것이 확실히 더 잘 읽힐 것 같은데, 프리코스 미션 중에는 다음 input을 받기 전에 예외를 throw하고 해당 스텝부터 다시 input을 받으라는 요구사항이 있었어서 고민이 됐습니다 😵‍💫

하지만 그렇다고 input 검증 부분을 뷰에서 온전히 처리하도록 하고 싶어도, 뷰가 모르는 내부 비즈니스 로직에서의 예외 상황도 있을 수 있기에 좋은 코드와 요구사항을 모두 만족하는 이상적인 상황을 어떻게 만들어나가면 좋을지를 의식적으로 생각해보려고 합니다..!

ddaaac commented on src/main/java/domain/AttendanceRecord.java:
```diff
@@ -0,0 +1,87 @@
+        this.attendance = Attendance.getAttendanceStatus(Day.getDay(dateTime.toLocalDate()), dateTime.toLocalTime());
+    }
+
+    public static AttendanceRecord parse(String dateTime) {
+        return new AttendanceRecord(LocalDateTime.parse(dateTime.replace(" ", "T")));
+    }
```
String -> Record로의 변환 책임까지 이 객체가 가져야할까에 대해서도 한 번 고민해보시면 좋을 것 같습니다.

jbilee replied:
지금 드는 생각은 `AttendanceRecord`는 LocalDate를 인자로 받아 저장하기만 하는게 역할에 더 적합할 것 같습니다. 이 부분은 2단계에서 한번 시도해보겠습니다!

ddaaac reviewed (approved):
안녕하세요 줄리 👋 

리뷰 반영 잘 해주셔서 특별히 더 남길 코멘트가 없네요.

2단계 진행해주세요.
2단계 진행하면서는 도메인을 모르는 상태(1단계)에서의 TDD와 도메인 파악이 어느 정도 된 상태에서의 TDD가 어떻게 다른지 느껴보시면 좋을 것 같습니다.
